### PR TITLE
[IMP] mrp: search bom by reference

### PR DESCRIPTION
**Pull request by yhu-odoo**
When select a BOM in the field, now we only search by the finished
product name. In this commit, we will also match the reference of the
BOM.

Task 2326786

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_bom.py [hunk 1]
```diff
@@ -145,6 +145,15 @@ def unlink(self):
             raise UserError(_('You can not delete a Bill of Material with running manufacturing orders.\nPlease close or cancel it first.'))
         return super(MrpBom, self).unlink()
 
+    @api.model
+    def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
+        args = args or []
+        if operator == 'ilike' and not (name or '').strip():
```
Comment (amoyaux) [inline addons/mrp/models/mrp_bom.py:151]: What about '=', '=ilike', 'like', '=like' and all the opposite (not ilike...)

Comment (yhu-odoo) [reply addons/mrp/models/mrp_bom.py]: @amoyaux fixed it

File: addons/mrp/models/mrp_bom.py [hunk 2]
```diff
@@ -145,6 +145,16 @@ def unlink(self):
             raise UserError(_('You can not delete a Bill of Material with running manufacturing orders.\nPlease close or cancel it first.'))
         return super(MrpBom, self).unlink()
 
+    @api.model
+    def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
+        args = args or []
+        positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
```
Comment (amoyaux) [inline addons/mrp/models/mrp_bom.py:151]: You forgot negative opperator

Comment (yhu-odoo) [reply addons/mrp/models/mrp_bom.py]: @amoyaux Should be ok now

File: addons/mrp/models/mrp_bom.py [hunk 3]
```diff
@@ -145,6 +146,16 @@ def unlink(self):
...
         return super(MrpBom, self).unlink()
 
+    @api.model
+    def _name_search(self, name='', args=None, operator='ilike', limit=100, name_get_uid=None):
+        args = args or []
+        domain = []
+        if (name or '').strip():
+            domain = ['|', (self._rec_name, operator, name), ('code', operator, name)]
+            if operator in expression.NEGATIVE_TERM_OPERATORS:
+                domain = ['&', '!'] + domain[1:]
```
Comment (amoyaux) [inline addons/mrp/models/mrp_bom.py:156]: You could use distribute_not

File: addons/mrp/models/mrp_bom.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
 from odoo import api, fields, models, _
 from odoo.exceptions import UserError, ValidationError
+from odoo.osv import expression
```
Comment (amoyaux) [inline addons/mrp/models/mrp_bom.py:6]: from odoo.osv.expression import distribute_not, NEGATIVE_TERM_OPERATORS

**Comment by amoyaux** [comment]
robodoo r+

**Pull request closed without merging**